File: Modules/BuiltInAlgorithms/Sources/IntroSort.swift
import AlgorithmKit
import Foundation
import SortEngineKit

public struct IntroSort: SortAlgorithm {
    public let id = AlgorithmID(rawValue: "introsort")
    public let metadata = AlgorithmMetadata(
        displayName: "Intro Sort",
        category: .logarithmic,
        sizeRange: 16...512,
        stable: false,
        timeComplexity: ComplexityBounds(best: "O(n log n)", average: "O(n log n)", worst: "O(n log n)"),
        spaceComplexity: "O(log n)",
        iconName: "square.stack.3d.up"
    )

    public init() {}

    public func record(into engine: inout RecordingEngine) {
        let n = engine.count
        let sizeThreshold = 16

        func medianOf3(_ left: Int, _ mid: Int, _ right: Int) -> Int {
            if !engine.compare(right, left) {
                engine.swap(left, right)
            }
            if !engine.compare(mid, left) {
                engine.swap(mid, left)
            }
            if !engine.compare(right, mid) {
                engine.swap(right, mid)
            }
            return mid
        }

        func partition(_ lo: Int, _ hi: Int, _ pivotValue: Int) -> Int {
            var i = lo, j = hi
            while true {
                while engine.values[i] < pivotValue { i += 1 }
                j -= 1
                while pivotValue < engine.values[j] { j -= 1 }
                if !(i < j) { return i }
                engine.swap(i, j)
                i += 1
            }
        }

        func heapSortRange(_ lo: Int, _ hi: Int) {
            let size = hi - lo

            func siftDown(_ root: Int, _ rangeSize: Int) {
                var root = root
                while true {
                    var largest = root
                    let left = 2 * root + 1
                    let right = 2 * root + 2
                    if left < rangeSize && !engine.compare(lo + largest, lo + left) { largest = left }
                    if right < rangeSize && !engine.compare(lo + largest, lo + right) { largest = right }
                    if largest == root { break }
                    engine.swap(lo + root, lo + largest)
                    root = largest
                }
            }

            var i = size / 2 - 1
            while i >= 0 {
                siftDown(i, size)
                i -= 1
            }
            var end = size - 1
            while end > 0 {
                engine.swap(lo, lo + end)
                siftDown(0, end)
                end -= 1
            }
        }

        func introsortLoop(_ lo: Int, _ hi: Int, _ depthLimit: Int) {
            var hi = hi
            var depthLimit = depthLimit
            while hi - lo > sizeThreshold {
                if depthLimit == 0 {
                    heapSortRange(lo, hi)
                    return
                }
                depthLimit -= 1
                let mid = lo + (hi - lo) / 2
                let pivotIndex = medianOf3(lo, mid, hi - 1)
                let pivotValue = engine.values[pivotIndex]
                let p = partition(lo, hi, pivotValue)
                introsortLoop(p, hi, depthLimit)
                hi = p
            }
        }

        func insertionSort(_ start: Int, _ end: Int) {
            var i = start + 1
            while i < end {
                var j = i
                while j > start && !engine.compare(j, j - 1) {
                    engine.swap(j - 1, j)
                    j -= 1
                }
                i += 1
            }
        }

        func floorLog2(_ a: Int) -> Int {
            Int((log(Double(a)) / log(2.0)).rounded(.down))
        }

        introsortLoop(0, n, 2 * floorLog2(n))
        insertionSort(0, n)
    }
}

File: App/Resources/AlgorithmDetails/introsort.bundle/introsort.cpp
#include <cstdio>
#include <cmath>
#include <utility>

int array[16] = {0, 39, 21, 62, 91, 77, 14, 23,
                 90, 69, 51, 81, 68, 83, 32, 56};

void printList(int items[], int size) {
  for (int i = 0; i < size; i++) {
    if (i == 0) {
      printf("[%d, ", items[i]);
    } else if (i != size - 1) {
      printf("%d, ", items[i]);
    } else {
      printf("%d]", items[i]);
    }
  }
}

int medianOf3(int arr[], int left, int mid, int right) {
  if (!(arr[left] >= arr[right])) {
    std::swap(arr[left], arr[right]);
  }
  if (!(arr[left] >= arr[mid])) {
    std::swap(arr[left], arr[mid]);
  }
  if (!(arr[mid] >= arr[right])) {
    std::swap(arr[mid], arr[right]);
  }
  return mid;
}

int partition(int arr[], int lo, int hi, int pivotValue) {
  int i = lo, j = hi;
  while (true) {
    while (arr[i] < pivotValue) i++;
    j--;
    while (pivotValue < arr[j]) j--;
    if (!(i < j)) return i;
    std::swap(arr[i], arr[j]);
    i++;
  }
}

void siftDown(int arr[], int lo, int root, int rangeSize) {
  while (true) {
    int largest = root;
    int left = 2 * root + 1;
    int right = 2 * root + 2;
    if (left < rangeSize && arr[lo + largest] < arr[lo + left]) largest = left;
    if (right < rangeSize && arr[lo + largest] < arr[lo + right]) largest = right;
    if (largest == root) break;
    std::swap(arr[lo + root], arr[lo + largest]);
    root = largest;
  }
}

void heapSortRange(int arr[], int lo, int hi) {
  int size = hi - lo;
  for (int i = size / 2 - 1; i >= 0; i--) siftDown(arr, lo, i, size);
  for (int end = size - 1; end > 0; end--) {
    std::swap(arr[lo], arr[lo + end]);
    siftDown(arr, lo, 0, end);
  }
}

void insertionSort(int arr[], int start, int end) {
  for (int i = start + 1; i < end; i++) {
    int j = i;
    while (j > start && arr[j] < arr[j - 1]) {
      std::swap(arr[j - 1], arr[j]);
      j--;
    }
  }
}

int floorLog2(int a) {
  return static_cast<int>(std::floor(std::log(static_cast<double>(a)) / std::log(2.0)));
}

void introsortLoop(int arr[], int lo, int hi, int depthLimit) {
  while (hi - lo > 16) {
    if (depthLimit == 0) {
      heapSortRange(arr, lo, hi);
      return;
    }
    depthLimit--;
    int mid = lo + (hi - lo) / 2;
    int pivotIndex = medianOf3(arr, lo, mid, hi - 1);
    int pivotValue = arr[pivotIndex];
    int p = partition(arr, lo, hi, pivotValue);
    introsortLoop(arr, p, hi, depthLimit);
    hi = p;
  }
}

void sort(int arr[], int n) {
  introsortLoop(arr, 0, n, 2 * floorLog2(n));
  insertionSort(arr, 0, n);
}

int main(int argc, char *argv[]) {
  int size = sizeof(array) / sizeof(array[0]);
  sort(array, size);
  printList(array, size);
  return 0;
}
